File: app/src/main/java/com/botijasoftware/utils/Vector3.java
package com.botijasoftware.utils;

public class Vector3 {
	
	public Vector3() {
		X = Y = Z = 0.0f;
	}

	public Vector3(float val) {
		X = val;
		Y = val;
		Z = val;
	}
	public Vector3(float valx, float valy, float valz) {
		X = valx;
		Y = valy;
		Z = valz;
	}
	
	public void setValue(float val) {
		X = val;
		Y = val;
		Z = val;
	}
	
	public void setValue(float valx, float valy, float valz) {
		X = valx;
		Y = valy;
		Z = valz;
	}
	
	public void setValue(Vector3 v) {
		X = v.X;
		Y = v.Y;
		Z = v.Z;
	}
	
	public Vector3 clone() {
		return new Vector3(X,Y,Z);
	}
	
	public Vector3 normalize() {
		float len = length();
		if (len != 0.0f) {
			X /= len;
			Y /= len;
			Z /= len;
		}
		return this;
	}
	
	public Vector3 add(Vector3 v) {
		X += v.X;
		Y += v.Y;
		Z += v.Z;
		return this;
	}
	
	public Vector3 add(float val) {
		X += val;
		Y += val;
		Z += val;
		return this;
	}
	
	public Vector3 sub(Vector3 v) {
		X -= v.X;
		Y -= v.Y;
		Z -= v.Z;
		return this;
	}
	
	public Vector3 sub(float val) {
		X -= val;
		Y -= val;
		Z -= val;
		return this;
	}
	
	public Vector3 mul(Vector3 v) {
		X *= v.X;
		Y *= v.Y;
		Z *= v.Z;
		return this;
	}
	
	public Vector3 mul(float val) {
		X *= val;
		Y *= val;
		Z *= val;
		return this;
	}
	
	public Vector3 div(Vector3 v) {
		if (v.X != 0.0f && v.Y != 0.0f && v.Z != 0.0f){
			X /= v.X;
			Y /= v.Y;
			Z /= v.Z;
		}
		return this;
	}
	
	public Vector3 div(float val) {
		if (val != 0.0f){
			float inv = 1.0f/val;
			X *= inv;
			Y *= inv;
			Z *= inv;
		}
		return this;
	}
	
	public float lengthsq() {
		return X * X + Y * Y + Z * Z;
	}
	
	public float length() {
		return (float)Math.sqrt( X * X + Y * Y + Z * Z);
	}
	
	public boolean isZero() {
		return (X <= SIGMA && X >= -SIGMA && Y <= SIGMA && Y >= -SIGMA && Z <= SIGMA && Z >= -SIGMA);
	}
	
	public boolean equals(float val) {
		float val_plus = val + SIGMA;
		float val_minus = val - SIGMA;
		return (X <= val_plus && X >= val_minus && Y <= val_plus && Y >= val_minus && Z <= val_plus && Z >= val_minus);
	}
	
	public boolean equals(Vector3 v) {
		return (X <= v.X+SIGMA && X >= v.X-SIGMA && Y <= v.Y+SIGMA && Y >= v.Y-SIGMA && Z <= v.Z+SIGMA && Z >= v.Z-SIGMA);
	}
	
	public float distancesq(Vector3 v) {
		return ((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z));
	}
	
	public float distance(Vector3 v) {
		return (float)Math.sqrt( (X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z) );
	}
	
	public float dot(Vector3 v) {
		return X * v.X + Y * v.Y + Z * v.Z;
	}
	
	public Vector3 cross(Vector3 v) {
		float nx = Y * v.Z - Z * v.Y;
		float ny = Z * v.X - X * v.Z;
		float nz = X * v.Y - Y * v.X;
		X = nx;
		Y = ny;
		Z = nz;
		return this;
	}
	
	public Vector3 reflect(Vector3 v) {
		//i - (2 * n * dot(i, n))
		sub(v.mul( 2.0f * dot(v)));
		return this;
	}

	
	public Vector3 lerp(Vector3 v, float factor) {
		if (factor <=0.0f) {
			return this;
		}
		else if (factor >= 1.0f) {
			X = v.X;
			Y = v.Y;
			Z = v.Z;
			return this;
		}
		
		X = X + factor * ( v.X - X);
		Y = Y + factor * ( v.Y - Y);
		Z = Z + factor * ( v.Z - Z);
		
		return this;
	}
	
	
	public Vector3 zero() {
		X = 0.0f;
		Y = 0.0f;
		Z = 0.0f;
		return this;
	}
	
	public static Vector3 cross(Vector3 v, Vector3 u) {
		float nx = v.Y * u.Z - v.Z * u.Y;
		float ny = v.Z * u.X - v.X * u.Z;
		float nz = v.X * u.Y - v.Y * u.X;		
		return new Vector3(nx, ny, nz);
	}
	
	public static Vector3 orthoNormalize(Vector3 n, Vector3 u) {
		n.normalize();
		Vector3 v = cross(n,u);
		v.normalize();
		return cross(v,n);		
	}
	
	public static Vector3 lerp(Vector3 v, Vector3 u, float factor) {
		if (factor <=0.0f) {
			return v.clone();
		}
		else if (factor >= 1.0f) {
			return u.clone();
		}
		
		float x = v.X + factor * ( u.X - v.X);
		float y = v.Y + factor * ( u.Y - v.Y);
		float z = v.Z + factor * ( u.Z - v.Z);
		
		return new Vector3(x,y,z);
	}
	
	public float X;
	public float Y;
	public float Z;
	private final static float SIGMA = 0.000001f;
	public static final Vector3 ZERO = new Vector3(0.0f, 0.0f, 0.0f);
	public static final Vector3 ONE = new Vector3(1.0f, 1.0f, 1.0f);
    public static final Vector3 RIGHT = new Vector3(1.0f, 0.0f, 0.0f);
    public static final Vector3 UP = new Vector3(0.0f, 1.0f, 0.0f);
    public static final Vector3 FORWARD = new Vector3(0.0f, 0.0f, 1.0f);
}

File: app/src/main/java/com/botijasoftware/utils/Vector3.kt
package com.botijasoftware.utils

class Vector3 {

    constructor() {
        Z = 0.0f
        Y = Z
        X = Y
    }

    constructor(value: Float) {
        X = value
        Y = value
        Z = value
    }

    constructor(valx: Float, valy: Float, valz: Float) {
        X = valx
        Y = valy
        Z = valz
    }

    fun setValue(value: Float) {
        X = value
        Y = value
        Z = value
    }

    fun setValue(valx: Float, valy: Float, valz: Float) {
        X = valx
        Y = valy
        Z = valz
    }

    fun setValue(v: Vector3) {
        X = v.X
        Y = v.Y
        Z = v.Z
    }

    fun clone(): Vector3 {
        return Vector3(X, Y, Z)
    }

    fun normalize(): Vector3 {
        val len = length()
        if (len != 0.0f) {
            X /= len
            Y /= len
            Z /= len
        }
        return this
    }

    fun add(v: Vector3): Vector3 {
        X += v.X
        Y += v.Y
        Z += v.Z
        return this
    }

    fun add(value: Float): Vector3 {
        X += value
        Y += value
        Z += value
        return this
    }

    fun sub(v: Vector3): Vector3 {
        X -= v.X
        Y -= v.Y
        Z -= v.Z
        return this
    }

    fun sub(value: Float): Vector3 {
        X -= value
        Y -= value
        Z -= value
        return this
    }

    fun mul(v: Vector3): Vector3 {
        X *= v.X
        Y *= v.Y
        Z *= v.Z
        return this
    }

    fun mul(value: Float): Vector3 {
        X *= value
        Y *= value
        Z *= value
        return this
    }

    operator fun div(v: Vector3): Vector3 {
        if (v.X != 0.0f && v.Y != 0.0f && v.Z != 0.0f) {
            X /= v.X
            Y /= v.Y
            Z /= v.Z
        }
        return this
    }

    operator fun div(value: Float): Vector3 {
        if (value != 0.0f) {
            val inv = 1.0f / value
            X *= inv
            Y *= inv
            Z *= inv
        }
        return this
    }

    fun lengthsq(): Float {
        return X * X + Y * Y + Z * Z
    }

    fun length(): Float {
        return Math.sqrt((X * X + Y * Y + Z * Z).toDouble()).toFloat()
    }

    val isZero: Boolean
        get() = X <= SIGMA && X >= -SIGMA && Y <= SIGMA && Y >= -SIGMA && Z <= SIGMA && Z >= -SIGMA

    fun equals(value: Float): Boolean {
        val val_plus = value + SIGMA
        val val_minus = value - SIGMA
        return X in val_minus..val_plus && Y <= val_plus && Y >= val_minus && Z <= val_plus && Z >= val_minus
    }

    fun equals(v: Vector3): Boolean {
        return X <= v.X + SIGMA && X >= v.X - SIGMA && Y <= v.Y + SIGMA && Y >= v.Y - SIGMA && Z <= v.Z + SIGMA && Z >= v.Z - SIGMA
    }

    fun distancesq(v: Vector3): Float {
        return (X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z)
    }

    fun distance(v: Vector3): Float {
        return Math.sqrt(((X - v.X) * (X - v.X) + (Y - v.Y) * (Y - v.Y) + (Z - v.Z) * (Z - v.Z)).toDouble()).toFloat()
    }

    fun dot(v: Vector3): Float {
        return X * v.X + Y * v.Y + Z * v.Z
    }

    fun cross(v: Vector3): Vector3 {
        val nx = Y * v.Z - Z * v.Y
        val ny = Z * v.X - X * v.Z
        val nz = X * v.Y - Y * v.X
        X = nx
        Y = ny
        Z = nz
        return this
    }

    fun reflect(v: Vector3): Vector3 {
        //i - (2 * n * dot(i, n))
        sub(v.mul(2.0f * dot(v)))
        return this
    }


    fun lerp(v: Vector3, factor: Float): Vector3 {
        if (factor <= 0.0f) {
            return this
        } else if (factor >= 1.0f) {
            X = v.X
            Y = v.Y
            Z = v.Z
            return this
        }

        X += factor * (v.X - X)
        Y += factor * (v.Y - Y)
        Z += factor * (v.Z - Z)

        return this
    }


    fun zero(): Vector3 {
        X = 0.0f
        Y = 0.0f
        Z = 0.0f
        return this
    }

    var X: Float = 0.0f
    var Y: Float = 0.0f
    var Z: Float = 0.0f

    companion object {

        fun cross(v: Vector3, u: Vector3): Vector3 {
            val nx = v.Y * u.Z - v.Z * u.Y
            val ny = v.Z * u.X - v.X * u.Z
            val nz = v.X * u.Y - v.Y * u.X
            return Vector3(nx, ny, nz)
        }

        fun orthoNormalize(n: Vector3, u: Vector3): Vector3 {
            n.normalize()
            val v = cross(n, u)
            v.normalize()
            return cross(v, n)
        }

        fun lerp(v: Vector3, u: Vector3, factor: Float): Vector3 {
            if (factor <= 0.0f) {
                return v.clone()
            } else if (factor >= 1.0f) {
                return u.clone()
            }

            val x = v.X + factor * (u.X - v.X)
            val y = v.Y + factor * (u.Y - v.Y)
            val z = v.Z + factor * (u.Z - v.Z)

            return Vector3(x, y, z)
        }

        private val SIGMA = 0.000001f
        val ZERO = Vector3(0.0f, 0.0f, 0.0f)
        val ONE = Vector3(1.0f, 1.0f, 1.0f)
        val RIGHT = Vector3(1.0f, 0.0f, 0.0f)
        val UP = Vector3(0.0f, 1.0f, 0.0f)
        val FORWARD = Vector3(0.0f, 0.0f, 1.0f)
    }
}
